### PR TITLE
Classification

### PR DESCRIPTION
**Pull request by lexmapr**
Revamping of classification component based on newer IFSAC+ classific…
…ation schema.

Includes updated predefined resources.
Addition of some new predefined resources.
Functionality to deal with multi class label values in bucket labels.
Functionality to customize ordering of class labels in the final class assignment.
Updated and newer rules for post-classification refinement of  IFSAC labels.
Excluded "hasBroadSynonym" from the mapping considerations
Functionality to convert the matched component to standard Upper case Ontology Ids
Four new modules for assignment of confidence level in mapping (for future use)

### CODE REVIEW

File: lexmapr/pipeline_classification.py [hunk 1]
```diff
@@ -2,84 +2,222 @@
...
+                                "environmental-abattoir", "environmental-warehouse",
+                                "environmental-researchfacility",
+                                "environmental-pasture", "environmental-animal housing",
+                                "environmental-factory/production facility",
+                                "environmental-vehicle", "environmental-construction"}
 
+    revised_final_labels_list = []
+    ifsac_final_labels_list = list(ifsac_final_labels)
+    if ret.intersection(priority_listing_categories):
+        priority_category =  (ret.intersection(priority_listing_categories)).pop()
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:29]:
```suggestion
        priority_category = (ret.intersection(priority_listing_categories)).pop()
```

File: lexmapr/pipeline_classification.py [hunk 2]
```diff
@@ -2,84 +2,222 @@
...
+            for multi_label in multi_labels:
+                revised_final_labels.add(multi_label)
+        else:
+            revised_final_labels.add(label)
+    revised_final_labels_list = list(revised_final_labels)
+    return revised_final_labels_list
+
+
+def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
+    """Get the refined final labels after application of customized rules.
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:56]: Should be 72 chars or less.

File: lexmapr/pipeline_classification.py [hunk 3]
```diff
@@ -2,84 +2,222 @@
...
+    :param str sample: sample
+    :param set ifsac_final_labels: the final labels set
+    :param dict label_refinements: the dictionary of label refinement resource
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:59]:
```suggestion
    :param dict label_refinements: the dictionary of label refinement 
        resource
```

File: lexmapr/pipeline_classification.py [hunk 4]
```diff
@@ -2,84 +2,222 @@
...
+    :rtype: set
+    """
+    # Caution: Rules are sequential - changing the order might change results
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:62]:
```suggestion
    # Caution: Rules are sequential - changing the order might change 
    # results.
```
Comments should be not exceed 72 chars per line.

File: lexmapr/pipeline_classification.py [hunk 5]
```diff
@@ -2,84 +2,222 @@
...
+    ret = set(ifsac_final_labels)
+    sample =  sample = helpers.punctuation_treatment(sample)
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:64]:
```suggestion
    sample = helpers.punctuation_treatment(sample)
```

File: lexmapr/pipeline_classification.py [hunk 6]
```diff
@@ -2,84 +2,222 @@
...
     sample_tokens = word_tokenize(sample)
-    sample_tokens = list(map(lambda token: singularize(token), sample_tokens))
+    sample_tokens_set = set(sample_tokens)
+
     for label, refined_label in label_refinements.items():
         label_tokens = word_tokenize(label)
-        label_tokens = list(map(lambda token: singularize(token), label_tokens))
-        if not (set(label_tokens) - set(sample_tokens)):
+        #label_tokens = list(map(lambda token: singularize(token), label_tokens))
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:70]:
```suggestion
```

File: lexmapr/pipeline_classification.py [hunk 7]
```diff
@@ -2,84 +2,222 @@
...
+        if not (set(label_tokens) - set(sample_tokens)) or re.search (r"\b"+label+r"\b",sample):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:71]:
```suggestion
        if not (set(label_tokens) - set(sample_tokens)) or re.search(r"\b"+label+r"\b", sample):
```

File: lexmapr/pipeline_classification.py [hunk 8]
```diff
@@ -2,84 +2,222 @@
...
+    aquatic_animal_categories = {"fish", "other aquatic animals"} | shellfish_categories
+    poultry_categories = {"other poultry", "chicken", "turkey"}
+    avian_categories = {"other poultry", "game", "poultry"} | poultry_categories
+
+    animal_categories = {"human",  "companion animal", "aquatic animals", "wild animal",
+                         "beef", "pork", "other meat", "cow", "pig"}
+    animal_categories |= avian_categories | aquatic_animal_categories | {"other animal"}
+    veterinary_categories = avian_categories | aquatic_animal_categories | {"other animal"}
+    veterinary_categories |= {"animal", "avian", "companion animal", "aquatic animals", "wild animal",
+                                "beef", "pork", "other meat", "cow", "pig"}
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:88]:
```suggestion
    veterinary_categories |= {"animal", "avian", "companion animal", "aquatic animals", 
                              "wild animal", "beef", "pork", "other meat", "cow", "pig"}
```

File: lexmapr/pipeline_classification.py [hunk 9]
```diff
@@ -2,84 +2,222 @@
...
+                                "environmental-abattoir", "environmental-warehouse",
+                                "environmental-researchfacility",
+                                "environmental-pasture", "environmental-animal housing",
+                                "environmental-factory/production facility",
+                                "environmental-vehicle", "environmental-construction"}
+    root_underground_categories = {"root/underground (roots)", "root/underground (tubers)",
+                                   "root/underground (bulbs)", "root/underground (other)"}
+    seeded_vegetable_categories = {"seeded vegetables (vine-grown)",
+                                   "seeded vegetables (solanaceous)", "seeded vegetables (legumes)",
+                                   "seeded vegetables (other)"}
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:100]:
```suggestion
    seeded_vegetable_categories = {"seeded vegetables (vine-grown)",
                                   "seeded vegetables (solanaceous)", 
                                   "seeded vegetables (legumes)",
                                   "seeded vegetables (other)"}
```

File: lexmapr/pipeline_classification.py [hunk 10]
```diff
@@ -2,84 +2,222 @@
...
+                            "vegetable row crops (leafy)"}
+    vegetable_categories |= root_underground_categories | seeded_vegetable_categories
+    fruit_categories = {"melon fruit", "pome fruit", "stone fruit", "sub-tropical fruit",
+                        "small fruit", "tropical fruit"}
+    plant_categories = {"oils", "vegetables", "fruits", "grains", "beans", "nuts",
+                        "seeds"}
+    plant_categories |= vegetable_categories | fruit_categories
+
+    other_plant_food_category = {"other (food additive)", "dietary supplement", "other (sweetener)",
+                                 "other (flavoring and seasoning", "other (confectionary)"}
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:112]:
```suggestion
    other_plant_food_category = {"other (food additive)", "dietary supplement", 
                                 "other (sweetener)", "other (flavoring and seasoning", 
                                 "other (confectionary)"}
```

File: lexmapr/pipeline_classification.py [hunk 11]
```diff
@@ -2,84 +2,222 @@
...
+    other_animal_food_category = {"meat", "other meat", "beef", "pork"}
+
+    # Customized rules for refinement of class labels
+    # Deals with "animal feed" class
+    if "animal feed" in ret and "by" in sample and not "by product" in sample:
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:117]:
```suggestion
    if "animal feed" in ret and "by" in sample and "by product" not in sample:
```

File: lexmapr/pipeline_classification.py [hunk 12]
```diff
@@ -2,84 +2,222 @@
...
+        ret.remove("animal feed")
+
+    # Deals with "clinical/research" class
+    if "clinical/research" in ret and ret.intersection((plant_categories
+            | other_plant_food_category)) and not ("swab" in sample or "clinical" in sample):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:122]:
```suggestion
    if "clinical/research" in ret \
        and ret.intersection(plant_categories | other_plant_food_category) \
        and not ("swab" in sample or "clinical" in sample):
```

File: lexmapr/pipeline_classification.py [hunk 13]
```diff
@@ -2,84 +2,222 @@
...
+        ret.remove("clinical/research")
+    if "clinical/research" in ret  and "swab sub" in sample:
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:124]:
```suggestion
    if "clinical/research" in ret and "swab sub" in sample:
```

File: lexmapr/pipeline_classification.py [hunk 14]
```diff
@@ -2,84 +2,222 @@
...
+        ret.clear()
         ret.add("environmental")
+    if "clinical/research" in ret and "scat" in sample:
+        ret.remove("clinical/research")
+        ret.add("environmental")
+    if "clinical/research" in ret and "environmental" in ret \
+            and not ("tissue" in sample or "biologicl" in sample):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:131]: Does this work? `"biologicl"` is misspelled.

File: lexmapr/pipeline_classification.py [hunk 15]
```diff
@@ -2,84 +2,222 @@
...
+        ret.remove("clinical/research")
+    if "clinical/research" in ret  and ret.intersection(environmental_categories):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:133]:
```suggestion
    if "clinical/research" in ret and ret.intersection(environmental_categories):
```

File: lexmapr/pipeline_classification.py [hunk 16]
```diff
@@ -2,84 +2,222 @@
...
+        ret.remove("clinical/research")
+    if "clinical/research" in ret and (ret.intersection(plant_categories)
+                                       or ret.intersection(animal_categories)):
+        if ("shell" in sample or  "shell on" in sample or "shellon" in sample):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:137]:
```suggestion
        if "shell" in sample or "shell on" in sample or "shellon" in sample:
```

File: lexmapr/pipeline_classification.py [hunk 17]
```diff
@@ -2,84 +2,222 @@
...
+
+    if "animal" in ret and sample != "animal":
+        ret.remove("animal")
+        ret.add("other animal")
+
+    # Dealing with "dairy", "cow" and "beef" cases
     if "dairy" in ret and "cow" in ret:
         ret.remove("cow")
-    if "beef" in ret and "dairy" in ret and "milk" in sample:
+    if "beef" in ret and "dairy" in ret and "milk" in sample :
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:152]:
```suggestion
    if "beef" in ret and "dairy" in ret and "milk" in sample:
```

File: lexmapr/pipeline_classification.py [hunk 18]
```diff
@@ -2,84 +2,222 @@
...
+        ret.remove("beef")
+    beef_keywords = ["raw cow", "raw veal", "raw calf", "meat", "beef",
+                     "cow lung", "cow liver", "cow heart"]
+    for entry in beef_keywords:
+        if entry in sample and "cow" in ret:
+            ret.remove("cow")
+            ret.add("beef")
+    pork_keywords = ["raw pig", "raw swine", "meat", "pork", "porcine"]
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:160]: `pork_keywords` is never used.

File: lexmapr/pipeline_classification.py [hunk 19]
```diff
@@ -2,84 +2,222 @@
...
+    for entry in beef_keywords:
+        if entry in sample and "pig" in ret:
+            ret.remove("pig")
+            ret.add("pork")
+    if "cow" in ret and "beef" in ret:  #REMOVE
+        ret.remove("cow")
+    if "beef" in ret and "fish" in ret  and ("fillet" in sample or "filet" in sample):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:167]:
```suggestion
    if "beef" in ret and "fish" in ret and ("fillet" in sample or "filet" in sample):
```

File: lexmapr/pipeline_classification.py [hunk 20]
```diff
@@ -2,84 +2,222 @@
...
         ret.remove("beef")
-    if "beef" in ret and ("cow" in ret or "calf" in ret):
+    if "beef" in ret and ("veterinary clinical/research" in ret ):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:169]:
```suggestion
    if "beef" in ret and ("veterinary clinical/research" in ret):
```

File: lexmapr/pipeline_classification.py [hunk 21]
```diff
@@ -2,84 +2,222 @@
...
-    # # #
+    if "fish" in ret and "eggs" in ret:
+        ret.remove("eggs")
+    if "fish eggs" in ret and "eggs" in ret:
+        ret.remove("fish eggs")
+    if "fish" in ret and "poultry" in ret:
+        ret.remove("poultry")
+    if "fish" in ret and "other poultry" in ret:
+        ret.remove("other poultry")
+    if ("poultry" in ret and "eggs" in ret):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:188]:
```suggestion
    if "poultry" in ret and "eggs" in ret:
```

File: lexmapr/pipeline_classification.py [hunk 22]
```diff
@@ -2,84 +2,222 @@
...
+    if "pork" in ret and "veterinary clinical/research" in ret:
         ret.remove("pork")
         ret.add("pig")
-    if "beef" in ret and ("clinical/research" in ret or "dairy" in ret):
-        ret.remove("beef")
-        ret.add("cow")
-    if "other meat" in ret and "clinical/research" in ret:
+    if "meat" in ret and ("veterinary clinical/research" in ret or "engineering  seafood" in ret):
+        ret.remove("meat")
+    if ret.intersection(specific_meat_categories) and "meat" in ret:  #sample has also to be considered
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:203]:
```suggestion
    if ret.intersection(specific_meat_categories) and "meat" in ret:
```

File: lexmapr/pipeline_classification.py [hunk 23]
```diff
@@ -2,84 +2,222 @@
...
-    aquatic_animal_categories = {"fish", "other aquatic animals"} | shellfish_categories
-
-    poultry_categories = {"other poultry", "chicken", "turkey"}
-    avian_categories = {"game", "poultry"} | poultry_categories
-
-    animal_categories = {"human", "avian", "companion animal", "aquatic animals", "wild animal",
-                         "beef", "pork", "other meat", "cow", "pig", "other animal"}
-    animal_categories |= avian_categories | aquatic_animal_categories
 
+    # Retains the specific (more granular) animal classes
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:220]:
```suggestion
    # Retains the specific (more granular) animal classes
```

File: lexmapr/pipeline_classification.py [hunk 24]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
     if "plant" in ret and ret.intersection(plant_categories):
         ret.remove("plant")
 
-    if "food" in ret and ret.intersection(animal_categories | plant_categories | {"meat"}):
-            ret.remove("food")
+    if "nut" in ret and "seeds" in ret and len(ret) == 2:
+         ret.remove("seeds")
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:252]:
```suggestion
        ret.remove("seeds")
```

File: lexmapr/pipeline_classification.py [hunk 25]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+    if "environment-water" in ret and "fish" in ret and len(ret) == 2:
+         ret.remove("environment-water")
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:254]:
```suggestion
        ret.remove("environment-water")
```

File: lexmapr/pipeline_classification.py [hunk 26]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+    if ("environmental-animal housing" in ret or "environmental-abattoir" in ret
+            or "environmental-farm" in ret) \
+            and "environmental-factory/production facility" in ret:
+        ret.remove("environmental-factory/production facility")
+    if "environmental-abattoir" in ret and "environmental-factory/production facility" in ret:
+        ret.remove("environmental-factory/production facility")
+
+    if not ("clinical/research" in ret or "veterinary clinical/research" in ret or "animal feed" in ret
+            or "human" in ret or "environmental" in ret or ret.intersection(environmental_categories)) \
+            and len(ret) >= 3:
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:268]:
```suggestion
    exclusions = {
        'clinical/research', 'veterinary clinical/research', 'animal feed', 'human',
        'environmental'
    }
    if not (ret.intersection(exclusions) or ret.intersection(environmental_categories)) \
            and len(ret) >= 3:
```

Line is too long, and kind of messy. You can do things like this to clean it up.

File: lexmapr/pipeline_classification.py [hunk 27]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+        ret.add("multi-ingredient")   # **********
+
+    if "other meat" in ret and "other animal" in ret:
+         ret.remove("other animal")
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:272]:
```suggestion
        ret.remove("other animal")
```

File: lexmapr/pipeline_classification.py [hunk 28]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+    if "meat" in ret and ret.intersection(animal_categories):
+        if len(ret) == 3 and "multi-ingredient" in ret :
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:274]:
```suggestion
        if len(ret) == 3 and "multi-ingredient" in ret:
```

File: lexmapr/pipeline_classification.py [hunk 29]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+            ret.remove("multi-ingredient")
+            ret.remove("meat")
+        else:
+            ret.remove("meat")
+
+    # Retains the specific (more granular) classes and removing the general "food" class
+    if "food" in ret and ret.intersection(animal_categories | plant_categories |
+                            other_animal_food_category| other_plant_food_category |{"plant","animal"}):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:282]:
```suggestion
    if "food" in ret and ret.intersection(animal_categories | plant_categories 
                                          | other_animal_food_category | other_plant_food_category 
                                          | {"plant", "animal"}):
```

File: lexmapr/pipeline_classification.py [hunk 30]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+        ret.remove("food")
+    if "food" in ret and ("dairy" in ret or "environmental" in ret or
+                        "clinical/research" in ret or "veterinary clinical/research" in ret):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:285]:
```suggestion
    if "food" in ret and ("dairy" in ret or "environmental" in ret or "clinical/research" in ret 
                          or "veterinary clinical/research" in ret):
```

File: lexmapr/pipeline_classification.py [hunk 31]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+    if "environmental-animal housing" in ret and "finished" in sample:
+        ret.remove("environmental-animal housing")
+    if ("chicken" in ret or "poultry" in ret or "other poultry" in ret or "cow" in ret) \
+            and "environmental-factory/production facility" in ret:
+        ret.remove("environmental-factory/production facility")
+        ret.add("environmental-farm")
+    if "eggs" in ret and "veterinary clinical/research" in ret:
+        ret.remove("veterinary clinical/research")
+    if "environmental" in ret and ("multi-ingredient" in ret or ret.intersection(plant_categories))\
+            and not ("swab" in sample or "environmental" in sample):
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:300]:
```suggestion
    if "environmental" in ret \
        and ("multi-ingredient" in ret or ret.intersection(plant_categories)) 
        and not ("swab" in sample or "environmental" in sample):
```

File: lexmapr/pipeline_classification.py [hunk 32]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+        ret.remove("environmental")
+
+    food_anatomical_parts = {'heart','liver','lung', 'leg', 'shell-on','shell', 'soft shell', 'tail','hlso','shellon',
+                             'beef','pork','meat','porcine','shell on'}
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:304]:
```suggestion
    food_anatomical_parts = {'heart', 'liver', 'lung', 'leg', 'shell-on', 'shell', 'soft shell',
                             'tail', 'hlso', 'shellon', 'beef', 'pork', 'meat', 'porcine',
                             'shell on'}
```

File: lexmapr/pipeline_classification.py [hunk 33]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+
+    body_part_for_food_animal_categories = aquatic_animal_categories | shellfish_categories \
+                                           | poultry_categories |{"cow"}
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:307]:
```suggestion
    body_part_for_food_animal_categories = \
        aquatic_animal_categories | shellfish_categories | poultry_categories | {"cow"}
```

File: lexmapr/pipeline_classification.py [hunk 34]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+    if "veterinary clinical/research" in ret and ret.intersection(body_part_for_food_animal_categories)\
+        and sample_tokens_set.intersection(food_anatomical_parts) and not "swab" in sample:
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:309]:
```suggestion
    if "veterinary clinical/research" in ret \
            and ret.intersection(body_part_for_food_animal_categories) \
            and sample_tokens_set.intersection(food_anatomical_parts) and "swab" not in sample:
```

File: lexmapr/pipeline_classification.py [hunk 35]
```diff
@@ -133,14 +248,80 @@ def refine_ifsac_final_labels(sample, ifsac_final_labels, label_refinements):
...
+        ret.remove("veterinary clinical/research")
+
+    disambiguation_words = {'ground', 'scraps', 'cut', 'smoke', 'moon', 'plain'}
+    if "environmental" in ret and (ret.intersection(animal_categories) or ret.intersection(plant_categories) or "dairy" in ret)\
+            and sample_tokens_set.intersection(disambiguation_words) :
```
Comment (ivansg44) [inline lexmapr/pipeline_classification.py:314]:
```suggestion
    if "environmental" in ret \
            and (ret.intersection(animal_categories) or ret.intersection(plant_categories) 
                 or "dairy" in ret) \
            and sample_tokens_set.intersection(disambiguation_words):
```

File: lexmapr/pipeline.py [hunk 36]
```diff
@@ -127,15 +139,21 @@ def run(args):
...
             cleaned_sample = helpers.abbreviation_normalization_phrase(cleaned_sample,
                                                                        lookup_table, micro_status)
             cleaned_sample = helpers.non_English_normalization_phrase(cleaned_sample, lookup_table,
                                                                       micro_status)
+            cleaned_sample_scientific_name = helpers.get_annotated_sample(cleaned_sample_scientific_name,
+                                                                               lemma, scientific_names_dict)
+            cleaned_sample_scientific_name = re.sub(' +', ' ', cleaned_sample_scientific_name)
 
-        cleaned_sample = helpers.remove_duplicate_tokens(cleaned_sample)
+        if "gallus gallus" not in sample:  # not to remove duplicate from sample containing gallus gallus
```
[the remark this answers is not in the document]
Comment (lexmapr) [reply lexmapr/pipeline.py]: Done.

**Comment by lexmapr** [comment]
A new pull request has been ctreated with all the requested changes and additional functionalities.

**Pull request closed without merging**